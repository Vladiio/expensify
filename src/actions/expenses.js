import uuid from 'uuid';
import './constants';

export const editExpense = (
  id,
  updates
) => ({
  type: EDIT_EXPENSE,
  id,
  updates
});

export const addExpense = ({
  description = '',
  note = '',
  amount = 0,
  createdAt = 0
} = {}) => ({
  type: ADD_EXPENSE,
  id: uuid(),
  description,
  note,
  amount,
  createdAt
});

export const removeExpense = ({
  id
} = {}) => ({
  type: REMOVE_EXPENSE,
  id
});