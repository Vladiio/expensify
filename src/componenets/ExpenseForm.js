import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);
    const { expense } = props;
    this.state = {
      description: expense ? expense.description : '',
      note: expense ? expense.note : '',
      amount: expense ? expense.amount / 100 : '',
      createdAt: expense ? moment(expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onSubmitForm = event => {
    event.preventDefault();
    const { description, amount } = this.state;
    let error = '';

    if (!description || !amount) {
      error = 'Please provide description and amount';
    }
    this.setState({ error });
    if (error) {
      return;
    }

    this.props.onSubmit({
      description,
      amount: parseFloat(amount, 10) * 100,
      note: this.state.note,
      createdAt: this.state.createdAt.valueOf()
    });
  };

  onInputChange = event => {
    const fieldValue = event.target.value;
    const fieldName = event.target.name;
    if (
      fieldName === 'amount' &&
      (fieldValue && !fieldValue.match(/^\d{1,}(\.\d{0,2})?$/))
    ) {
      return;
    }

    this.setState({
      [fieldName]: fieldValue
    });
  };

  onDateChange = createdAt => {
    createdAt &&
      this.setState({
        createdAt
      });
  };

  onFocusChange = ({ focused }) => {
    this.setState({
      calendarFocused: focused
    });
  };

  render() {
    return (
      <div>
        {this.state.error && <h3>{this.state.error}</h3>}
        ExpenseForm
        <form onSubmit={this.onSubmitForm}>
          <input
            type="text"
            placeholder="description"
            autoFocus
            name="description"
            value={this.state.description}
            onChange={this.onInputChange}
          />
          <input
            type="text"
            name="amount"
            value={this.state.amount}
            onChange={this.onInputChange}
            placeholder="Amount"
          />{' '}
          <br />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            name="note"
            value={this.state.note}
            onChange={this.onInputChange}
            placeholder="Add a note for your expense (optional)"
          />
          <button>Add expense</button>
        </form>
      </div>
    );
  }
}
