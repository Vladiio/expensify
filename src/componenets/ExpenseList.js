import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = (props) => (
  <div>
    <h3>Expense List</h3>
    <ul>
    {
      props.expenses.map(
        expense => <ExpenseListItem key={expense.id} {...expense}/>
      )
    }
    </ul>
  </div>
);

const mapStateToProps = ({
  expenses,
  filter
}) => {
  return {

    expenses: selectExpenses(expenses, filter)
  }
};

export default connect(
  mapStateToProps
)(ExpenseList);