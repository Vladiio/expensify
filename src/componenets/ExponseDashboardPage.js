import React from 'react';

const ExpenseDashboardPage = () => (
  <div>
    This is from my dashboard componenet
  </div>
);

export default ExpenseDashboardPage;