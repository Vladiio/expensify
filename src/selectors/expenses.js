import moment from 'moment';

const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  const filteredExpenses = expenses.filter(
    ({ description, createdAt }) => {
      const createdAtMoment = moment(createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      const textMatch = description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    }
  );

  return filteredExpenses.sort((expenseOne, expenseTwo) => {
    switch (sortBy) {
      case 'date':
        return expenseTwo.createdAt - expenseOne.createdAt;

      case 'amount':
        return expenseTwo.amount - expenseOne.amount;
      default:
        return 0;
    }
  });
};

export default getVisibleExpenses;
