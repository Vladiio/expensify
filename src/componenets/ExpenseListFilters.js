import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setFilter, sortByAmount, sortByDate, setEndDate, setStartDate } from '../actions/filters';
import 'react-dates/lib/css/_datepicker.css';



class ExpenseListFilters extends Component {

  state = {
    focusedInput: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  onFocusChange = (focusedInput) => {
    this.setState({ focusedInput });
  };

  render() {
    return (
      <div>
        <input 
        type="text"
          value={this.props.filter.text}
          onChange={(e) => this.props.onFilterChange(e.target.value)}/>
        <select
          value={this.props.filter.sortBy} 
          onChange={(e) => {this.props.onOrderChange(e.target.value)}}>
          <option defaultValue="date">date</option>
          <option defaultValue="amount">amount</option>
        </select>

        <DateRangePicker 
          startDate={this.props.filter.startDate}
          endDate={this.props.filter.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          startDateId='startDate1'
          endDateId='enddate2'
          showClearDates={true}
        />
      </div>
    );

  }
}

const mapStateToProps = ({ filter }) => ({
  filter
});

const mapDispatchToProps = (dispatch) => ({
  onFilterChange: (text) => {
    dispatch(setFilter(text))
  },
  onOrderChange: (orderBy) => {
    switch (orderBy) {
      case 'amount':
        dispatch(sortByAmount());
        break;
      default:
        dispatch(sortByDate());
    }
  },
  dispatch
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);