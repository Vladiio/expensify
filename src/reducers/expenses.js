import { ADD_EXPENSE, REMOVE_EXPENSE, EDIT_EXPENSE } from '../actions/constants';

const expensesReducerInitial = [];

const expensesReducer = (
  state = expensesReducerInitial,
  action
) => {
  switch (action.type) {
    case ADD_EXPENSE:
      return [
        ...state,
        {
          description: action.description,
          id: action.id,
          note: action.note,
          amount: action.amount,
          createdAt: action.createdAt
        }
      ];
    
    case REMOVE_EXPENSE:
      return state.filter(expense => (
        expense.id !== action.id
      ));
    case EDIT_EXPENSE:
      const { updates, id } = action;
      return state.map(expense => {
        if (expense.id === id) {
          return {
            ...expense,
            ...updates
          };
        return expense
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;