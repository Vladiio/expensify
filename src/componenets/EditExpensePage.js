import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';

import { removeExpense, editExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onFormSubmit = updates => {
    const { editExpense, expense } = this.props;
    editExpense(expense.id, updates);
  };

  onButtonClick = () => {
    const { removeExpense, expense, history } = this.props;
    removeExpense(expense.id);
    history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onFormSubmit}
        />
        <button onClick={this.onButtonClick}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  expense: state.expenses.find(
    expense => expense.id === ownProps.match.params.id
  )
});

const mapDispatchToProps = dispatch => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: id => dispatch(removeExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(
  EditExpensePage
);
