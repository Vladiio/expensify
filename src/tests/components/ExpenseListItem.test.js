import React from 'react';
import { shallow } from 'enzyme';

import expenses from '../fixtures/expenses';
import { ExpenseListItem } from '../../componenets/ExpenseListItem';

it('should render ExpenseListItem correctly', () => {
  const data = {
    id: '1',
    description: 'rent',
    amount: 500,
    createdAt: 5000
  };
  const wrapper = shallow(<ExpenseListItem {...data} />);
  expect(wrapper).toMatchSnapshot();
});
