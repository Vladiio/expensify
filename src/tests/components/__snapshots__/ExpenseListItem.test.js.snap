// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h4>
      rent
    </h4>
  </Link>
  <p>
    500
     - 
    5000
  </p>
</div>
`;
